projects/driver_controls/src/ads1015.c

```diff
@@ -0,0 +1,151 @@
+#include <ads1015.h>
+#include <stdbool.h>
+#include <stdio.h>
+
+#include "ads1015_bitmasks.h"
+#include "delay.h"
+#include "gpio_it.h"
+#include "i2c.h"
+
+// Union definition to help process registers data
+typedef struct ADS1015Register {
+  uint8_t data[2];
+} ADS1015Register;
+
+// Since the ADS1015 stores data in two's complement, an int16_t will be used to hold
+// the raw readings
+typedef struct ADS1015Interrupt {
+  ADS1015Callback callback;
+  void *context;
+  int16_t reading;
+} ADS1015Interrupt;
+
+static I2CPort s_i2c_port;
+static ADS1015Channel s_current_channel;
+
+static ADS1015Interrupt s_interrupts[NUM_ADS1015_CHANNELS];
+
+static StatusCode prv_channel_valid(ADS1015Channel channel) {
+  if (channel >= NUM_ADS1015_CHANNELS) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  }
+
+  return STATUS_CODE_OK;
+}
+
+// Initiates a read as outlined in section 8.5.3 of the datasheet
+static StatusCode prv_read(I2CPort i2c_port, uint8_t reg, uint8_t *data, size_t len) {
+  status_ok_or_return(i2c_write(i2c_port, ADS1015_I2C_ADDRESS, &reg, 1));
+  status_ok_or_return(i2c_read(i2c_port, ADS1015_I2C_ADDRESS, data, len));
+
+  return STATUS_CODE_OK;
+}
+
+// Initiates a write as outlined in section 8.5.3 of the datasheet
+static StatusCode prv_write(I2CPort i2c_port, uint8_t reg, uint8_t *data, size_t len) {
+  // For writes, the input data must be sent in the same frame as the address pointer register
+  uint8_t write_data[] = { reg, data[0], data[1] };
+  status_ok_or_return(i2c_write(i2c_port, ADS1015_I2C_ADDRESS, write_data, len));
+  return STATUS_CODE_OK;
+}
+
+static void prv_interrupt_handler(const GPIOAddress *address, void *context) {
+  ADS1015Register reg;
+
+  // Obtain ADC readings
+  prv_read(s_i2c_port, ADS1015_CONVERSION_REGISTER, reg.data, 2);
+
+  s_interrupts[s_current_channel].reading = (reg.data[0] << 4) | (reg.data[1] >> 4);
+
+  if (s_interrupts[s_current_channel].callback != NULL) {
+    s_interrupts[s_current_channel].callback(s_current_channel,
+                                             s_interrupts[s_current_channel].context);
+  }
+
+  // Update so that the ADC reads from the next channel
+  s_current_channel = (s_current_channel + 1) % NUM_ADS1015_CHANNELS;
+
+  // Obtain the current value for the config register
+  prv_read(s_i2c_port, ADS1015_CONFIG_REGISTER, reg.data, 2);
+
+  // Set bits 12-14 based on the current channel
+  reg.data[0] = ADS1015_CONFIG_MUX(reg.data[0], s_current_channel);
+  prv_write(s_i2c_port, ADS1015_CONFIG_REGISTER, reg.data, 3);
```

Comment:
Magic number? (I'll stop commenting on every occurrence) I'm really confused what is happening with this `len` field if `reg.data` is always 2 bytes.

Reply:
I removed the len parameter from the read and writes. The same number of bytes are sent for each transaction, so that magic number paramater can be removed